### PR TITLE
Script to separate the ligands added to scripts within asapdiscovery-data.

### PR DESCRIPTION
**Pull request by SiyangPeng**
## Description
Provide a brief description of the PR's purpose here.

## Todos
Notable points that this PR has either accomplished or will accomplish.
- [ ] Point 1 ...

## Questions
- [ ] Question 1 ..

## Status
- [ ] Ready to go

## Developers certificate of origin
- [ ] I certify that this contribution is covered by the MIT license as defined in our [LICENSE](https://github.com/choderalab/asapdiscovery/blob/main/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

### CODE REVIEW

File: asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+
+# The function to isolate ligands in original file and separate out individual ligands
+# Each ligand saved in individual pdb file
+def split_ligands(directory, input_file):
+    # Where original file is from
+    local_path = Path(directory)
+    # Where ending strucutures will be put into
+    os.mkdir(directory + "output/")
+    output_path = Path(directory + "output/")
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:49]: You can just do 

```suggestion
    output_path = Path(directory) / "output"
    output_path.mkdir()
```
and then you can get rid of the `os.mkdir` on L48

File: asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    # Load structure into pymol
+    cmd.load(local_path / input_file, "complex")
+
+    # Function that saves residues named "LIG" to a PDB file
+    def save_ligand_residues(structure_name, output_file):
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:55]: Would you be able to un-nest these nested functions? Its no big deal but un-nested is the norm.

File: asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                    G.add_edge(atom_numbers[0], atom1)
+    # Delete previous pymol structure before proceeding so no confusion
+    cmd.delete("complex")
+    # Load only ligands into Pymol
+    cmd.load(ligands, "ligands")
+    # Remove this pdb since no longer in use
+    os.remove(ligands)
+
+    # Function to save each group connected atoms (a ligand) into individual files (with bonds)
+    def save_subset_with_conect(structure_name, atom_ids, output_file):
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:85]: Same here for nested functions

File: asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    for lig_pdb_file in local_path.glob("lig_*.pdb"):
+        # Load in the ligand pdb into openeye
+        lig = load_openeye_pdb(lig_pdb_file)
+        # Put it back
+        lig_protein = combine_protein_ligand(protein, lig)
+        # Save in file
+        output_file = str(lig_pdb_file.stem) + "_protein.pdb"
+        save_openeye_pdb(lig_protein, output_path / output_file)
+        # Delete the ligand pdb since no longer useful
+        os.remove(lig_pdb_file)
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:136]: you can use `tempfile` if you want but no stress either way.

File: asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        # Save the selected subset with CONECT records to a PDB file
+        save_subset_with_conect("ligands", atom_ids, local_path / f"lig_{i}.pdb")
+
+    # Load pdb files and save all singled out ligands to singular combined sdf
+    ligs = []
+    for lig_pdb_file in local_path.glob("lig_*.pdb"):
+        lig = load_openeye_pdb(lig_pdb_file)
+        lig.SetTitle(lig_pdb_file.stem)
+        ligs.append(lig)
+    save_openeye_sdfs(ligs, output_path / "Combined_ligs.sdf")
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:112]:
```suggestion
    save_openeye_sdfs(ligs, output_path / "combined_ligs.sdf")
```

File: asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    # Function that saves residues named "LIG" to a PDB file
+    def save_ligand_residues(structure_name, output_file):
+        # Select residues named "LIG"
+        selection = f"{structure_name} and resn LIG"
+
+        # Save the selected residues to a PDB file
+        cmd.save(output_file, selection, format="pdb")
+
+    # Save all the ligands into one file
+    save_ligand_residues("complex", local_path / "ligand_only.pdb")
```
Comment (hmacdope) [inline asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:63]: Can you make `structure_name` (here "complex") into an arg (something like `args.c`) for more flexibility?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Looking great @SiyangPeng, just a few nitpicks. 

The other thing is could you add an endpoint in `asapdiscovery-data`'s `pyproject.toml? 

That way you can call it on the command line without needing `python` and it is added to environment on install. See here: https://github.com/choderalab/asapdiscovery/blob/main/asapdiscovery-data/pyproject.toml#L29

File: asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
+"""
+Script to convert a single pdb file with multiple ligands all named LIG in
```
[the remark this answers is not in the document]
Comment (apayne97) [reply asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:2]: what about 
`split_fragment_screen.py`
`process_fragment_screen_output.py`
`split_single_pdb_into_many_sdf_files.py`
`generate_single_ligand_complex_from_fragment_screen.py`

[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:2]: Happy with any of those.

[the remark this answers is not in the document]
Comment (SiyangPeng) [reply asapdiscovery-data/asapdiscovery/data/scripts/separate_ligand_script.py:2]: Great, will do.

**Review by hmacdope — APPROVED** [review]
LGTM @SiyangPeng! @apayne97 you happy with this?

**Review by hmacdope — CHANGES_REQUESTED** [review]
Whoops blocking for conflicts

**Comment by hmacdope** [comment]
@SiyangPeng, you will need to fix conflicts by merging with `upstream/main`, it may be easier to revert your (inconsequential) changes to the notebooks. Try running the following:

```
git fetch upstream
git merge upstream/main
... # make sure you fix conflicts
git commit
``

**Review by hmacdope — APPROVED** [review]
LGTM @SiyangPeng, thanks for this and congrats on your first contribution.

**Pull request merged**